### PR TITLE
FE-139: Support for search alias view type, replace views list with React

### PR DESCRIPTION
**Pull request by palashkaria**
### Scope & Purpose
- Replaces the views list & add view modal with React
- Adds a new view type - 'search-alias'
      - In Create modal
      - View screen - JSON editor
           -  save button enabled if changes are done
           - "copy mutable properties" shows only views of the same type

- [x] :pizza: New feature

### Checklist

- [x] Tests
  - [x] Manually tested
- [x] :book: CHANGELOG entry made

#### Related Information

- [x] GitHub issue / Jira ticket:  https://arangodb.atlassian.net/browse/FE-139
- [x] Design document:

### CODE REVIEW

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/viewsView.js
Comment (palashkaria) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/viewsView.js:551]: todo: remove hacky code

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/indicesView.js
Comment (palashkaria) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/indicesView.js:231]: todo: remove hacky code

**Review by adityamukho — COMMENTED** [review]
This is a partial review since the number of files is huge. Have left a few comments on some files.

I suggest splitting the PR into 2 - one for views lists and one for search alias views. The latter could also be raised only after the former is merged, to ensure inter-dependencies are maintained while still keeping a cap on the number of files per review.

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/ViewSettings.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/ViewSettings.tsx:104]: Yes, it still is. Logic is co-located the point of usage now, in the file `EditableViewNameField`

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/constants.ts
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/constants.ts:70]: No, that will actually cause issues. In the schema here, type is always going to be `arangosearch` as we don't want to change it from the JSON.

This type change is added here because we need to switch on top level at `ViewSettingsWrap`. It's a stop gap measure, as we need to refactor arangosearch stuff anyway.

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/forms/inputs/CopyFromInput.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/forms/inputs/CopyFromInput.tsx:26]: Sure, will do this.

In general - This field won't be too large, I don't think we need micro-optimizations on this level.
And even if we do hit a limit, this change would'nt be as impactful, as there are a lot of sub-optimal things happening in render all across all UIs. IMO we should fix these issues when we get them, and avoid pre-mature optimization.

**Comment by palashkaria** [comment]
> This is a partial review since the number of files is huge. Have left a few comments on some files.
> 
> I suggest splitting the PR into 2 - one for views lists and one for search alias views. The latter could also be raised only after the former is merged, to ensure inter-dependencies are maintained while still keeping a cap on the number of files per review.

This will be difficult to break into two PRs as the logic is very connected, and so is the task. It may take a long time to break into two PRs. Let me know how else can we make this easier to review, or if we want to a review in sync @adityamukho

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/useUpdateAliasViewProperties.ts
Comment (nikhil-varma) [inline js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/useUpdateAliasViewProperties.ts:127]: Maybe we should consider renaming this as follows
Instead of `changesAdded` to `addedChanges`

Similar for the deleted one. Let's follow the verb+entity/noun pattern

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/viewsList/ViewsList.tsx
Comment (nikhil-varma) [inline js/apps/system/_admin/aardvark/APP/react/src/views/views/viewsList/ViewsList.tsx:15]: Similar thing for this could also help. However, for later

**Review by nikhil-varma — CHANGES_REQUESTED** [review]

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/constants.ts
Comment (adityamukho) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/constants.ts:70]: You might want to take a look at `formStateSchema` (`views/forms/JsonForm.tsx: 35`). A number of immutable properties (set by user at view creation time) are set into the schema dynamically.

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/CopyPropertiesDropdown.tsx
Comment (adityamukho) [inline js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/CopyPropertiesDropdown.tsx:55]: Unnecessary indirection. Should be simplified to:
```
setOptions(newViews);
```

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/EditableViewNameField.tsx
Comment (adityamukho) [inline js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/EditableViewNameField.tsx:20]: Please extract the PropType.

**Review by adityamukho — COMMENTED** [review]
Left a few more comments. Haven't gone too deep into the search alias views, as per our discussion.

File: js/apps/system/_admin/aardvark/APP/react/src/components/form/InputControl.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/components/form/InputControl.tsx:10]: Same as above, I don't think it matters if it's default export or not. Sometimes named exports might be better. It sounds like a preference at this point, I wouldn't want to make change based on that

File: js/apps/system/_admin/aardvark/APP/react/src/components/form/CreatableMultiSelectControl.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/components/form/CreatableMultiSelectControl.tsx:18]: There's no real reason to prefer one over the other. You don't need to worry about what it is, the toolchain handles it.

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/constants.ts
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/constants.ts:70]: Yes, I am aware, and I have taken that into account. Those need no change, as this is only for arangosearch view. search-alias has a different way of doing things, and that can be ported over to arangosearch later, I'm not touching any of the arangosearch schema as of now.

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/viewsList/ViewsList.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/viewsList/ViewsList.tsx:14]: We can take this up in the future as we move more things to React. This is usually needed on the top level and can be at the layer of `<ChakraCustomProvider>`

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/EditableViewNameField.tsx
Comment (adityamukho) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/EditableViewNameField.tsx:22]: For readability. I didn't ask for it to be exported - just that it be declared as a type above and used here. However, this is your call.

File: js/apps/system/_admin/aardvark/APP/react/src/components/form/InputControl.tsx
Comment (adityamukho) [reply js/apps/system/_admin/aardvark/APP/react/src/components/form/InputControl.tsx:10]: It's not just my preference, but commonly used convention. Many well known libraries, including React and Lodash provide default exports where it makes sense (for example - in cases where the module name reflects its primary function).

File: js/apps/system/_admin/aardvark/APP/react/src/components/form/FormikFormControl.tsx
Comment (adityamukho) [reply js/apps/system/_admin/aardvark/APP/react/src/components/form/FormikFormControl.tsx:17]: Typescript may well handle this, just as it can handle any poorly written code as long as it compiles. My point was about this being more in line with conventions followed in most flagship modules, to make it easier for other developers to follow and use.

**Review by adityamukho — APPROVED** [review]
LGTM, in so far as it works, and I didn't notice any erroneous logic. However, for a PR this big, such issues could be easily missed. There are still a few points about convention, but those could be discussed, finalized and codified into a style guide and used for later PRs.

File: js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/EditableViewNameField.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/views/searchAliasView/EditableViewNameField.tsx:22]: I think readability is affected by consistency more than pulling things out. As we start using prettier everywhere, this becomes super easy to scan inline, instead of having devs jump around to definitions. Would prefer to stick to this convention (unless there's a need to reuse)

File: js/apps/system/_admin/aardvark/APP/react/src/components/form/InputControl.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/components/form/InputControl.tsx:10]: While these libraries do offer multiple different ways of importing, many of them exist for legacy reasons.

- A pitfall of using default exports is that it can be named anything in usage
- An advantage of using a named export is that it can allow IDEs to quickly suggest autocomplete

Common convention is not always the best, because the reasons behind them could be related to legacy code; not wanting breaking changes etc

**Pull request merged**
